book/src/rfcs/RFC0009-Updatable-Derived-Queries.md

```diff
@@ -0,0 +1,224 @@
+# Updatable derived queries
+
+## Metadata
+
+* Author: 1tgr
+* Date: 2021-06-17
+* Introduced in: https://github.com/salsa-rs/salsa/pull/269
+
+## Summary
+
+- Allow a query function to reuse the latest value on the same query
+
+## Motivation
+
+Derived queries in Salsa are implemented in terms of a pure function over a key and some other
+queries ('dependencies'). The value returned by the query is expected to depend only on the query
+key and the values of the dependencies.
+
+However, in performance terms, it can be helpful if the query's function can also see the most
+recent value calculated for the same query. This allows it to perform an incremental update to this
+existing value instead of constructing a new value from scratch.
+
+As a performance optimisation, this technique generally provides a benefit by allowing the query
+function to reuse an existing memory allocation ("mine it for spare parts"). Some examples:
+
+- Incrementally re-parsing a source file after a diff is applied
+- Trait matching, chalk-style, where we can generally keep the tables in between queries but if the
+  set of impls etc changes we may want to throw them away
+- Large values that want to reuse the allocation in a previous `Vec`, `String` or `HashMap`
+- Any persistent data structure
+
+## User's guide
+
+Conventionally, you define a Salsa query in terms of a single function on a trait:
+
+```rust
+#[salsa::query_group]
+trait MyDatabase {
+    fn formatted_value(&self, key: u32) -> String;
+}
+
+fn formatted_value(db: &dyn MyDatabase, key: u32) -> String {
+    db.some_input(key).to_string()
+}
+```
+
+When you apply the `#[salsa::update]` attribute to the function, Salsa looks for a second function,
+which it calls to update an existing value in response to changes in the query's dependencies. By
+default the update function has the word `update_` prepended to its name. Salsa passes an
+`&mut Value` reference to the update function, and it expects the update function to return `true`
+unless the value is unchanged.
+
+```rust
+#[salsa::query_group]
+trait MyDatabase {
+    #[salsa::update]
+    fn formatted_value(&self, key: u32) -> String;
+}
+
+fn formatted_value(db: &dyn MyDatabase, key: u32) -> String {
+    // Our query is being called for the first time and it has no value already cached.
+    // Return a new String.
+    db.some_input(key).to_string()
+}
+
+fn update_formatted_value(db: &dyn MyDatabase, value: &mut String, key: u32) -> bool {
```

Comment:
Nit: As noted elsewhere, I think the key(s) should come *first*. 

Other nit: I think I would prefer to return an enum like `salsa::ValueChanged::True` vs `salsa::ValueChanged::False`. We can also add a comment to that enum explaining that this result is **not a hint** and must be accurate, etc.

Reply:
I pushed this, although what about `ValueChanged::No` and `ValueChanged::Maybe`?